### PR TITLE
Refactoring: AST

### PR DESCRIPTION
**Pull request by aopoltorzhicky**
* Refactoring: parsing from procedures to typed trees
* NFT ledger
* Falphanet
* Metadata extra fields
* Fix many bugs

### CODE REVIEW

**Comment by aopoltorzhicky** [comment]
[Micheline AST](https://trello.com/c/aiwfzw5M/730-micheline-ast)

**Comment by aopoltorzhicky** [comment]
[Metadata: keep unknown fields](https://trello.com/c/EbBv2E3H/936-metadata-keep-unknown-fields)

**Comment by aopoltorzhicky** [comment]
[Operations: non-applied operation doesn't have displayed entrypoint](https://trello.com/c/HxvOu0B9/865-operations-non-applied-operation-doesnt-have-displayed-entrypoint)

**Comment by aopoltorzhicky** [comment]
[Florence](https://trello.com/c/ahWUXPht/961-florence)

**Comment by aopoltorzhicky** [comment]
[NFT ledger](https://trello.com/c/kXBiALx1/924-nft-ledger)

**Comment by aopoltorzhicky** [comment]
[Falphanet](https://trello.com/c/y2aiXted/931-falphanet)

**Comment by aopoltorzhicky** [comment]
[Florence [требует внимательной проверки]](https://trello.com/c/ahWUXPht/961-florence-%D1%82%D1%80%D0%B5%D0%B1%D1%83%D0%B5%D1%82-%D0%B2%D0%BD%D0%B8%D0%BC%D0%B0%D1%82%D0%B5%D0%BB%D1%8C%D0%BD%D0%BE%D0%B9-%D0%BF%D1%80%D0%BE%D0%B2%D0%B5%D1%80%D0%BA%D0%B8)

**Comment by aopoltorzhicky** [comment]
[400](https://trello.com/c/QOqh0gLe/776-400)

**Comment by aopoltorzhicky** [comment]
[Elastic real total](https://trello.com/c/dWojINJw/979-elastic-real-total)

**Comment by aopoltorzhicky** [comment]
[Tests: endpoints used by Atomex](https://trello.com/c/5geA15Qd/972-tests-endpoints-used-by-atomex)

File: internal/bcd/ast/forge.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package ast
+
+import "github.com/baking-bad/bcdhub/internal/bcd/forge"
+
+// Forge -
+func Forge(node Base, optimized bool) (string, error) {
+	baseAST, err := node.ToBaseNode(optimized)
+	if err != nil {
+		panic(err)
```
Comment (RomanSerikov) [inline internal/bcd/ast/forge.go:9]: do not panic

File: cmd/api/handlers/bigmap.go [hunk 2]
```diff
@@ -307,47 +330,43 @@ func (ctx *Context) prepareBigMapItem(data []bigmapdiff.BigMapDiff, keyHash stri
...
 			return nil, nil, "", err
 		}
-	}
-	var key interface{}
-	var keyString string
-	if item.Key != nil {
-		bKey, err := json.Marshal(item.Key)
+		key = keyMiguel
+
+		// TODO: unpack
```
Comment (RomanSerikov) [inline cmd/api/handlers/bigmap.go:342]: to-do

File: internal/bcd/ast/ast.go [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+	var s strings.Builder
+	for i := range a.Nodes {
+		s.WriteString(a.Nodes[i].String())
+	}
+	return s.String()
+}
+
+// Settle -
+func (a *TypedAst) Settle(untyped UntypedAST) error {
+	if len(untyped) != len(a.Nodes) && len(a.Nodes) == 1 {
```
Comment (RomanSerikov) [inline internal/bcd/ast/ast.go:81]: swap branches

File: internal/bcd/ast/baker_hash.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		Type:  strings.ToLower(s.Prim),
+		Value: value,
+		Name:  &name,
+	}, nil
+}
+
+// ToBaseNode -
+func (s *BakerHash) ToBaseNode(optimized bool) (*base.Node, error) {
+	val := s.Value.(string)
+	if optimized {
```
Comment (RomanSerikov) [inline internal/bcd/ast/baker_hash.go:50]: swap branches

File: internal/bcd/ast/code.go [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package ast
```
Comment (RomanSerikov) [inline internal/bcd/ast/code.go:1]: remove if useless

File: internal/bcd/ast/compare.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+			return 0, err
+		}
+		y.ValueKind = valueKindBytes
+		y.Value = value
+	}
+
+	return strings.Compare(x.Value.(string), y.Value.(string)), nil
+}
+
+func compareBigInt(x, y Default) int {
```
Comment (RomanSerikov) [inline internal/bcd/ast/compare.go:30]: add checks for bigint types

File: internal/bcd/ast/jsonschema.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	return nil
+}
+
+func setChildSchemaForMap(child Node, needXTitle bool, parent *JSONSchema) error {
+	s, err := child.ToJSONSchema()
+	if err != nil {
+		return err
+	}
+
+	if len(s.Properties) > 0 {
```
Comment (RomanSerikov) [inline internal/bcd/ast/jsonschema.go:162]: swap branches

File: internal/bcd/ast/list.go [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+	}
+	list.Type = typ
+
+	return nil
+}
+
+// ParseValue -
+func (list *List) ParseValue(node *base.Node) error {
+	if node.Prim != consts.PrimArray {
+		return errors.Wrap(consts.ErrInvalidPrim, "List.ParseValue")
```
Comment (RomanSerikov) [inline internal/bcd/ast/list.go:75]: add node.Prim or more node info to error (error context)

File: internal/bcd/ast/parameter.go [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package ast
```
Comment (RomanSerikov) [inline internal/bcd/ast/parameter.go:1]: remove if useless

File: internal/bcd/ast/storage.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package ast
```
Comment (RomanSerikov) [inline internal/bcd/ast/storage.go:1]: remove if useless

File: internal/bcd/ast/unit.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+func (u *Unit) ToParameters() ([]byte, error) {
+	return []byte(`{"prim":"Unit"}`), nil
+}
+
+// Compare -
+func (u *Unit) Compare(second Comparable) (int, error) {
+	if _, ok := second.(*Unit); !ok {
+		return 0, consts.ErrTypeIsNotComparable
+	}
+	return 0, nil
```
Comment (RomanSerikov) [inline internal/bcd/ast/unit.go:41]: no need to compare?

File: internal/bcd/ast/parameter.go (hunk 9 above)
Comment (aopoltorzhicky) [reply internal/bcd/ast/parameter.go:1]: Needed

File: internal/bcd/contract/strings.go [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+	sha := sha512.New()
+	if _, err := sha.Write(data); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(sha.Sum(nil)), nil
+}
+
+func findHardcodedAddresses(code []byte) types.Set {
+	regexString := "(tz|KT)[0-9A-Za-z]{34}"
+	re := regexp.MustCompile(regexString)
```
Comment (RomanSerikov) [inline internal/bcd/contract/strings.go:22]: should compile 1 time

File: internal/bcd/contract/strings.go [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+	res := re.FindAllString(string(code), -1)
+	resp := make(types.Set)
+	resp.Append(res...)
+	return resp
+}
+
+// IsAddress -
+func IsAddress(str string) bool {
+	regexString := "(tz|KT)[0-9A-Za-z]{34}"
+	re := regexp.MustCompile(regexString)
```
Comment (RomanSerikov) [inline internal/bcd/contract/strings.go:32]: should compile 1 time

File: internal/elastic/contract/storage.go [hunk 14]
```diff
@@ -349,7 +349,7 @@ func (storage *Storage) GetSameContracts(c contract.Contract, manager string, si
 					core.MatchPhrase("address", c.Address),
 				),
 			),
-		).Sort("last_action", "desc")
```
Comment (m-kus) [inline internal/elastic/contract/storage.go]: ??

Comment (aopoltorzhicky) [reply internal/elastic/contract/storage.go]: Посмотри, что за запрос, и станет понятно, что чтобы посчитать количество сортировать не надо

**Pull request merged**